src/test/java/net/minecraftforge/debug/NetworkCapabilityTest.java

```diff
@@ -0,0 +1,173 @@
+    });
+
+    public NetworkCapabilityTest()
+    {
+        FMLJavaModLoadingContext.get().getModEventBus().addListener(NetworkCapabilityTest::handleRegisterCapabilities);
+    }
+
+    private static void handleRegisterCapabilities(RegisterCapabilitiesEvent event)
+    {
+        event.register(TestCapability.class);
+    }
+
+    private static class TestCapability {
+
+        private int number;
+        private boolean dirty;
+
+        public void update()
+        {
+            this.number++;
+            this.dirty = true;
+        }
+    }
+
+    private static class TestCapabilityProvider implements ICapabilityProvider, INetworkCapability {
+
+        private final LazyOptional<TestCapability> cap = LazyOptional.of(() -> new TestCapability());
+
+        private final Supplier<String> name;
+
+        private TestCapabilityProvider(Supplier<String> name)
+        {
+            this.name = name;
+        }
+
+        @Override
+        public void writeCapabilities(FriendlyByteBuf out, boolean writeAll)
+        {
+            this.cap.ifPresent(cap ->
+            {
+                logger.info("[{}] Sending new number: {}", this.name.get(), cap.number);
+                cap.dirty = false;
+                out.writeVarInt(cap.number);
+            });
+        }
+
+        @Override
+        public void readCapabilities(FriendlyByteBuf in)
+        {
+            this.cap.ifPresent(cap ->
+            {
+                cap.number = in.readVarInt();
+                logger.info("[{}] Received new number: {}", this.name.get(), cap.number);
+            });
+        }
+
+        @Override
+        public boolean requiresSync()
+        {
+            return this.cap.map(cap -> cap.dirty).orElse(false);
+        }
+
+        @Override
+        public <T> LazyOptional<T> getCapability(Capability<T> cap, Direction side)
+        {
+            return cap == TEST_CAP ? this.cap.cast() : LazyOptional.empty();
+        }
+
+    }
+
+    @SubscribeEvent
+    public static void handleAttachEntityCapabilities(AttachCapabilitiesEvent<Entity> event)
+    {
+        event.addCapability(new ResourceLocation(ID, "test_cap"),
+            new TestCapabilityProvider(() -> event.getObject().getName().getString()));
+    }
+
+    @SubscribeEvent
+    public static void handleAttachItemStackCapabilities(AttachCapabilitiesEvent<ItemStack> event)
+    {
+        event.addCapability(new ResourceLocation(ID, "test_cap"),
+            new TestCapabilityProvider(() -> event.getObject().getDisplayName().getString()));
+    }
+
+    @SubscribeEvent
+    public static void handleAttachBlockEntityCapabilities(AttachCapabilitiesEvent<BlockEntity> event)
+    {
+        event.addCapability(new ResourceLocation(ID, "test_cap"),
+            new TestCapabilityProvider(() -> event.getObject().getBlockState().getBlock().getName().getString()));
+    }
+
+    @SubscribeEvent
+    public static void handleLivingEntityUseItem(LivingEntityUseItemEvent event)
```

Comment:
Bad test event, use one of the sub events. It gets difficult figuring out how the capability is updating when it could changed on start, tick, and stop.